### PR TITLE
v1 queues proposal

### PR DESCRIPTION
**Pull request by tjholm**
Restores queues from v0 with tweaks for v1

### CODE REVIEW

File: cloud/aws/deploy/policy.go [hunk 1]
```diff
@@ -86,6 +86,22 @@ var awsActionsMap map[resourcespb.Action][]string = map[resourcespb.Action][]str
...
+		"sqs:SendMessage",
+	},
+	resourcespb.Action_QueueReceive: {
+		"sqs:ReceiveMessage",
+		"sqs:DeleteMessage",
+	},
+	// XXX: Cannot be applied to single resources
+	// v1.Action_QueueList: {
+	// 	"sqs:ListQueues",
+	// },
```
Comment (jyecusch) [inline cloud/aws/deploy/policy.go:99]: Can this be deleted?

File: cloud/aws/deploy/queue.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"github.com/nitrictech/nitric/cloud/common/deploy/resources"
+	common "github.com/nitrictech/nitric/cloud/common/deploy/tags"
+	deploymentspb "github.com/nitrictech/nitric/core/pkg/proto/deployments/v1"
+	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/sqs"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+// Bucket - Implements deployments of Nitric Buckets using AWS S3
```
Comment (jyecusch) [inline cloud/aws/deploy/queue.go:27]:
```suggestion
// Queue - Implements deployments of Nitric Queues using AWS SQS
```

File: cloud/aws/runtime/queue/sqs.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	grpc_errors "github.com/nitrictech/nitric/core/pkg/grpc/errors"
+
+	queuepb "github.com/nitrictech/nitric/core/pkg/proto/queues/v1"
+)
+
+const (
+	// ErrCodeNoSuchTagSet - AWS API neglects to include a constant for this error code.
+	ErrCodeNoSuchTagSet = "NoSuchTagSet"
+	ErrCodeAccessDenied = "AccessDenied"
+)
```
Comment (jyecusch) [inline cloud/aws/runtime/queue/sqs.go:46]: Are these still used?

File: cloud/aws/runtime/queue/sqs.go [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+func (s *SQSQueueService) getUrlForQueueName(ctx context.Context, queue string) (*string, error) {
+	queues, err := s.provider.GetResources(ctx, resource.AwsResource_Queue)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving queue list")
+	}
+
+	queueArn, ok := queues[queue]
+
+	if !ok {
+		return nil, fmt.Errorf("queue %s does not exist", queue)
```
Comment (jyecusch) [inline cloud/aws/runtime/queue/sqs.go:65]:
```suggestion
		return nil, fmt.Errorf("arn for queue %s could not be determined", queue)
```

File: cloud/aws/runtime/queue/sqs.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+	client   sqsiface.SQSAPI
+}
+
+var _ queuepb.QueuesServer = &SQSQueueService{}
+
+// Get the URL for a given queue name
+func (s *SQSQueueService) getUrlForQueueName(ctx context.Context, queue string) (*string, error) {
+	queues, err := s.provider.GetResources(ctx, resource.AwsResource_Queue)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving queue list")
```
Comment (jyecusch) [inline cloud/aws/runtime/queue/sqs.go:59]:
```suggestion
		return nil, fmt.Errorf("error retrieving queue list: %w", err)
```

File: cloud/aws/runtime/queue/sqs.go [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+func isSQSAccessDeniedErr(err error) bool {
+	var opErr *smithy.OperationError
+	if errors.As(err, &opErr) {
+		return opErr.Service() == "SQS" && strings.Contains(opErr.Unwrap().Error(), "AccessDenied")
+	}
+	return false
+}
+
+func (s *SQSQueueService) Send(ctx context.Context, req *queuepb.QueueSendRequestBatch) (*queuepb.QueueSendResponse, error) {
```
Comment (jyecusch) [inline cloud/aws/runtime/queue/sqs.go:91]: Should we update the proto to only have a 'QueueSend' req/res, which can include multiple items and remove the batch operation names (or vice versa)?

File: cloud/aws/runtime/queue/sqs.go [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+				err,
+			)
+		}
+	} else {
+		return nil, newErr(
+			codes.NotFound,
+			"unable to find queue",
+			err,
+		)
+	}
```
Comment (jyecusch) [inline cloud/aws/runtime/queue/sqs.go:163]: Move err handling up so it's compacted by editors and reduces nesting?

```suggestion
	url, err := s.getUrlForQueueName(ctx, req.QueueName)
	if err != nil {
		return nil, newErr(codes.NotFound, "unable to find queue", err)
	}
	
	entries := make([]types.SendMessageBatchRequestEntry, 0)

	for _, sendTaskReq := range req.Requests {
		t := sendTaskReq

		// generate a unique Id for each task
		id := uuid.New()
		requestIdMap[id.String()] = t

		bytes, err := proto.Marshal(t.Payload)
		if err != nil {
			return nil, newErr(codes.Internal, "error marshalling task to JSON", err)
		}
		entries = append(entries, types.SendMessageBatchRequestEntry{
			Id:          aws.String(id.String()),
			MessageBody: aws.String(string(bytes)),
		})
	}

	out, err := s.client.SendMessageBatch(ctx, &sqs.SendMessageBatchInput{
		Entries:  entries,
		QueueUrl: url,
	})
	if err != nil {
		if isSQSAccessDeniedErr(err) {
			return nil, newErr(
				codes.PermissionDenied,
				"unable to send tasks to queue, have you requested access to this queue?",
				err,
			)
		}

		return nil, newErr(codes.Internal, "error sending tasks", err)
	}
	
	// process out failed messages to return to the caller
	failedTasks := make([]*queuepb.FailedSendRequest, 0, len(out.Failed))
	for _, failed := range out.Failed {
		for id, e := range requestIdMap {
			if id == *failed.Id {
				failedTasks = append(failedTasks, &queuepb.FailedSendRequest{
					Request: e,
					Message: *failed.Message,
				})
				// continue processing failed messages
				break
			}
		}
	}
	
	return &queuepb.QueueSendResponse{
		FailedRequests: failedTasks,
	}, nil
```

File: cloud/aws/runtime/queue/sqs.go [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+// Completes a previously popped queue item
```
Comment (jyecusch) [inline cloud/aws/runtime/queue/sqs.go:227]:
```suggestion
// Complete a previously received queue item, removing it from the queue
```

File: cloud/azure/deploy/queue.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+// 	res.Queue, err = storage.NewQueue(ctx, utils.ResourceName(ctx, name, utils.StorageQueueRT), &storage.QueueArgs{
+// 		AccountName:       args.Account.Name,
+// 		ResourceGroupName: args.ResourceGroup.Name,
+// 	})
+// 	if err != nil {
+// 		return nil, err
+// 	}
+
+// 	return res, nil
+// }
```
Comment (jyecusch) [inline cloud/azure/deploy/queue.go:80]: Delete?

File: cloud/azure/runtime/queue/azqueue.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+	"github.com/nitrictech/nitric/cloud/azure/runtime/env"
+	azqueueserviceiface "github.com/nitrictech/nitric/cloud/azure/runtime/queue/iface"
+	azureutils "github.com/nitrictech/nitric/cloud/azure/runtime/utils"
+	grpc_errors "github.com/nitrictech/nitric/core/pkg/grpc/errors"
+
+	queuespb "github.com/nitrictech/nitric/core/pkg/proto/queues/v1"
+)
+
+// Set to 30 seconds,
```
Comment (jyecusch) [inline cloud/azure/runtime/queue/azqueue.go:40]:
```suggestion
// Set to 30 seconds, a common value from other providers like AWS
```

File: cloud/azure/runtime/queue/azqueue.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+func (s *AzqueueQueueService) send(ctx context.Context, queueName string, req *queuespb.QueueSendRequest) (*queuespb.QueueSendResponse, error) {
+	newErr := grpc_errors.ErrorsWithScope("AzqueueQueueService.Send")
+
+	messages := s.getMessagesUrl(queueName)
+
+	// task := &azTask{
+	// 	Id:      req.Id,
+	// 	Payload: req.Payload.AsMap(),
+	// }
```
Comment (jyecusch) [inline cloud/azure/runtime/queue/azqueue.go:76]:
```suggestion
```

File: cloud/azure/runtime/queue/azqueue.go [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+func (s *AzqueueQueueService) getMessageIdUrl(queue string, messageId azqueue.MessageID) azqueueserviceiface.AzqueueMessageIdUrlIface {
+	mUrl := s.getMessagesUrl(queue)
+
+	return mUrl.NewMessageIDURL(messageId)
+}
+
+// type azTask struct {
+// 	Id      string
+// 	Payload map[string]interface{}
+// }
```
Comment (jyecusch) [inline cloud/azure/runtime/queue/azqueue.go:66]:
```suggestion
```

File: cloud/azure/runtime/queue/azqueue.go [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	return nil, nil
+}
+
+func (s *AzqueueQueueService) Send(ctx context.Context, req *queuespb.QueueSendRequestBatch) (*queuespb.QueueSendResponse, error) {
+	failedRequests := make([]*queuespb.FailedSendRequest, 0)
+
+	for _, r := range req.Requests {
+		// t := task
```
Comment (jyecusch) [inline cloud/azure/runtime/queue/azqueue.go:102]:
```suggestion
```

File: cloud/azure/runtime/queue/azqueue.go [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+		FailedRequests: failedRequests,
+	}, nil
+}
+
+// AzureQueueItemLease - Represents a lease of an Azure Storages Queues item
+// Azure requires a combination of their unique reference for a queue item (id) and a pop receipt (lease id)
+// to perform operations on the item, such as delete it from the queue.
+type AzureQueueItemLease struct {
+	// The ID of the queue item
+	// note: this is an id generated by Azure, it's not the user provided unique id.
```
Comment (jyecusch) [inline cloud/azure/runtime/queue/azqueue.go:123]:
```suggestion
	// note: this is an id generated by Azure
```

File: cloud/azure/runtime/queue/azqueue.go [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+			codes.Internal,
+			"failed to complete task",
+			err,
+		)
+	}
+
+	return &queuespb.QueueCompleteResponse{}, nil
+}
+
+const expiryBuffer = 2 * time.Minute
```
Comment (jyecusch) [inline cloud/azure/runtime/queue/azqueue.go:237]: Why is this 2 minutes?

File: cloud/gcp/deploy/queue.go [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+// 	Name         string
+// 	PubSub       *pubsub.Topic
+// 	Subscription *pubsub.Subscription
+// }
+
+// type PubSubTopicArgs struct {
+// 	Location string
+// 	StackID  string
+// 	Queue    *v1.Queue
+// }
```
Comment (jyecusch) [inline cloud/gcp/deploy/queue.go:42]:
```suggestion
```

File: cloud/gcp/runtime/queue/pubsub.go [hunk 17]
```diff
@@ -0,0 +1,359 @@
...
+	if err != nil {
+		return nil, newErr(
+			codes.NotFound,
+			"queue not found",
+			err,
+		)
+	}
+
+	// SendBatch once we've published all tasks to the client
+	// TODO: We may want to revisit this, and chunk up our publishing in a way that makes more sense...
```
Comment (jyecusch) [inline cloud/gcp/runtime/queue/pubsub.go:146]:
```suggestion
```

File: nitric/proto/queues/v1/queues.proto [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+option java_multiple_files = true;
+option java_outer_classname = "QueuesService";
+option php_namespace = "Nitric\\Proto\\Queues\\V1";
+option csharp_namespace = "Nitric.Proto.Queues.v1";
+
+// The Nitric Queue Service contract
+service Queues {
+  // Send a single event to a queue
+  rpc Send (QueueSendRequestBatch) returns (QueueSendResponse);
+  // Receive event(s) off a queue
```
Comment (jyecusch) [inline nitric/proto/queues/v1/queues.proto:18]:
```suggestion
  // Receive message(s) from a queue
```

File: nitric/proto/queues/v1/queues.proto [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+  rpc Receive (QueueReceiveRequest) returns (QueueReceiveResponse);
+  // Complete an event previously popped from a queue
```
Comment (jyecusch) [inline nitric/proto/queues/v1/queues.proto:20]:
```suggestion
  // Complete an item previously popped from a queue
```

**Review by jyecusch — CHANGES_REQUESTED** [review]

File: cloud/aws/deploy/policy.go (hunk 1 above)
Comment (tjholm) [reply cloud/aws/deploy/policy.go:99]: This was mostly to document out inability to apply that permission at the resource level, we have similiar comments for our other resources, did we want to clean them all out?

**Review by jyecusch — APPROVED** [review]

**Pull request merged**